generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -172,6 +173,74 @@ public OpaqueExpressionGenerator(IRandom generator, GenerationParams generationP
         depth, fuzzer)));
   }
 
+  /**
+   * Function to generate an opaque zero or one by taking the dot product of two vectors. This
+   * function depends on two rules of the dot product. Let u and v be vectors of the same size,
+   * and let i be a valid index into those vectors.
+   *     If u[i] != 0 <==> v[i] = 0, then dot(u, v) = 0. Example: u = (0, 1), v = (1, 0), then
+   *     dot(u, v) = 1 * 0 + 0 * 1 = 0.
+   *     If u[i] = v[i] = 0 for all i except for a single index, j, and if u[j] = v[j] = 1, then
+   *     dot(u, v) = 0. Example: u = (0, 1), v = (0, 1), then dot(u, v) = 0 * 0 + 1 * 1 = 1.
+   * Because fuzzed expressions are not always well defined, we forgo those in favor of simply
+   * using opaque ones and zeroes when generating vectors for this function.
+   * @param type - the base type of the opaque value being created.
+   * @param constContext - true if we're in a constant expression context, false otherwise.
+   * @param depth - how deep we are in the expression.
+   * @param fuzzer - the fuzzer object for generating fuzzed expressions.
+   * @param isZero - true if we are making an opaque zero, false otherwise.
+   * @return Optional.empty() if an opaque value can't be generated, otherwise an opaque value
+   *     made from the dot product of two vectors.
+   */
+  private Optional<Expr> opaqueZeroOrOneDot(BasicType type, boolean constContext,
+                                            final int depth, Fuzzer fuzzer, boolean isZero) {
+    if (type != BasicType.FLOAT) {
+      return Optional.empty();
+    }
+    // Vector sizes need to be in [2, 4].
+    final int vectorSize = generator.nextInt(3) + 2;
+    final Expr dotProductExpr;
+    if (isZero) {
+      // We're basically producing inverse vectors: u[i] = 0 <==> v[i] != 0.
+      // To do that, we need to store which values will be zero and which will not.
+      final List<Boolean> isZeroDeterminants = new ArrayList<Boolean>();
+      for (int i = 0; i < vectorSize; i++) {
```

Comment:
I don't know if there's some kind of initializer in the STL that lets you initialize a arbitrarily sized list with values - if there is, this would be a place to use it. (Or if there's a better way to do this than a list of booleans, let me know.)

Reply:
I would simplify the two loops to something like:

for (int i = 0; i < vectorSize; i++) {
  final boolean firstIsZero = generator.nextBoolean();
  firstVectorArgs.add( ..... use firstIsZero .... );
  secondVectorArgs.add( ..... use !firstIsZero .... );

Reply:
i.e. get rid of isZeroDeterminants.

Reply:
I don't know why I thought I couldn't inline that - whoops.